File: homework/src/main/java/com/mikheev/homework/controller/BookController.java
package com.mikheev.homework.controller;

import com.mikheev.homework.controller.dto.*;
import com.mikheev.homework.domain.Author;
import com.mikheev.homework.domain.Book;
import com.mikheev.homework.domain.Genre;
import com.mikheev.homework.service.AuthorService;
import com.mikheev.homework.service.BookService;
import com.mikheev.homework.service.GenreService;
import lombok.RequiredArgsConstructor;
import org.modelmapper.ModelMapper;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@RequiredArgsConstructor
@RestController
@RequestMapping("/api/library")
public class BookController {

    private final BookService bookService;
    private final AuthorService authorService;
    private final GenreService genreService;
    private final ModelMapper modelMapper;

    @GetMapping("/book")
    public List<BookListDto> bookList() {
        return bookService.getAllBooks().stream()
                .map(book -> modelMapper.map(book, BookListDto.class))
                .collect(Collectors.toList());
    }

    @GetMapping("/book/{id}")
    public BookDto getBook(@PathVariable("id") long id) {
        return modelMapper.map(bookService.getBookWithAllAssociations(id), BookDto.class);
    }

    @PreAuthorize("hasAnyRole('ADMIN', 'MODERATOR')")
    @GetMapping("/associations")
    public AssociationsDto loadEditCreateAssociations() {
        List<AuthorDto> authors = authorService.getAllAuthors().stream()
                .map(author -> modelMapper.map(author, AuthorDto.class)).collect(Collectors.toList());
        List<GenreDto> genres = genreService.getAllGenres().stream()
                .map(genre -> modelMapper.map(genre, GenreDto.class)).collect(Collectors.toList());
        return new AssociationsDto(authors, genres);
    }

    @PreAuthorize("hasAnyRole('ADMIN', 'MODERATOR')")
    @PostMapping("/book")
    public BookDto addBook(@RequestBody Book book) {
        return modelMapper.map(bookService.addBook(book), BookDto.class);
    }

    @PreAuthorize("hasAnyRole('ADMIN', 'MODERATOR')")
    @PutMapping("/book")
    public BookDto updateBook(@RequestBody Book book) {
        return modelMapper.map(bookService.updateBook(book), BookDto.class);
    }

    @PreAuthorize("hasAnyRole('ADMIN', 'MODERATOR')")
    @DeleteMapping("/book/{id}")
    public void deleteBook(@PathVariable("id") long id) {
        bookService.deleteBook(id);
    }
}

File: homework/src/main/kotlin/com/mikheev/homework/controller/BookController.kt
package com.mikheev.homework.controller

import com.mikheev.homework.controller.dto.*
import com.mikheev.homework.domain.Author
import com.mikheev.homework.domain.Book
import com.mikheev.homework.domain.Genre
import com.mikheev.homework.service.AuthorService
import com.mikheev.homework.service.BookService
import com.mikheev.homework.service.GenreService
import org.modelmapper.ModelMapper
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.util.stream.Collectors

@RestController
@RequestMapping("/api/library")
class BookController(
    val bookService: BookService,
    val authorService: AuthorService,
    val genreService: GenreService,
    val modelMapper: ModelMapper) {

    @GetMapping("/book")
    fun bookList(): List<BookListDto> {
        return bookService.getAllBooks().stream()
            .map { book: Book -> modelMapper.map(book, BookListDto::class.java) }
            .collect(Collectors.toList())
    }

    @GetMapping("/book/{id}")
    fun getBook(@PathVariable("id") id: Long): BookDto {
        return modelMapper.map(bookService.getBookWithAllAssociations(id), BookDto::class.java)
    }

    @GetMapping("/associations")
    fun loadEditCreateAssociations(): AssociationsDto {
        val authors = authorService.getAllAuthors().stream()
            .map { author: Author -> modelMapper.map(author, AuthorDto::class.java) }
            .collect(Collectors.toList())
        val genres = genreService.getAllGenres().stream()
            .map { genre: Genre -> modelMapper.map(genre, GenreDto::class.java) }
            .collect(Collectors.toList())
        return AssociationsDto(authors, genres)
    }

    @PostMapping("/book")
    fun addBook(@RequestBody book: Book): BookDto {
        return modelMapper.map(bookService.addBook(book), BookDto::class.java)
    }

    @PutMapping("/book")
    fun updateBook(@RequestBody book: Book): BookDto {
        return modelMapper.map(bookService.updateBook(book), BookDto::class.java)
    }

    @DeleteMapping("/book/{id}")
    fun deleteBook(@PathVariable("id") id: Long) {
        bookService.deleteBook(id)
    }
}
